pyuvdata/uvbeam.py

```diff
@@ -1102,54 +1156,56 @@ def _get_beam(self, pol):
-        # pols dict
-        pols_dict = {1: (-5, -6), 2: (-5, -6), 3: (-7, -8), 4: (-7, -8)}
-        # get pol array
         pol_array = self.polarization_array
-        # pQ, pU and pV yet to be implemented
-        allowed = [1, -5, -6, -7, -8]
-        if pol in allowed:
+        if pol in [-5, -6, -7, -8]:  # linear polarizations
             if pol in pol_array:
                 stokes_p_ind = np.where(np.isin(pol_array, pol))[0][0]
                 beam = self.data_array[0, 0, stokes_p_ind]
-            elif pol == 1:
-                keys = pols_dict[1]
-                if keys[0] in pol_array and keys[1] in pol_array:
-                    xx_ind = np.where(np.isin(pol_array, keys[0]))[0][0]
-                    yy_ind = np.where(np.isin(pol_array, keys[1]))[0][0]
-                    beam = 0.5 * (self.data_array[0, 0, xx_ind] + self.data_array[0, 0, yy_ind])
-                else:
-                    raise ValueError('Do not have the right polarization information')
             else:
                 raise ValueError('Do not have the right polarization information')
+        elif pol in [1, 2, 3, 4]:  # pseudo stokes
+            assert self.beam_type == 'efield', "Beam type must be efield for pseudo stokes {}".format(uvutils.polnum2str(pol))
+            # forming jones matrix
+            _sh = self.data_array.shape
+            jones = np.ndarray((_sh[-1], 2, 2), dtype=complex)
+            jones[:, 0, 0] = self.data_array[0, 0, 0, 0, :]
+            jones[:, 0, 1] = self.data_array[0, 0, 1, 0, :]
+            jones[:, 1, 0] = self.data_array[1, 0, 0, 0, :]
+            jones[:, 1, 1] = self.data_array[1, 0, 1, 0, :]
+            beam = self._construct_mueller(jones, pol - 1, pol - 1)
         else:
-            raise NotImplementedError("Polarization {} not yet implemented...".format(polstr))
+            raise ValueError('Do not have the right polarization information')
+
         return beam
 
     def get_beam_area(self, pol='pI'):
         """
         Computes the integral of the beam, which has units of steradians
 
-        Currently, only the pseudo-Stokes 'pI' beam and linear dipole 'XX' and 'YY' are
+        Pseudo-Stokes 'pI' (I), 'pQ'(Q), 'pU'(U), 'pV'(V) beam and linear dipole 'XX', 'XY', 'YX' and 'YY' are
         supported. See Equations 4 and 5 of Moore et al. (2017) ApJ 836, 154
-        or arxiv:1502.05072 for details.
+        or arxiv:1502.05072 and Kohn et al. (2018) or https://arxiv.org/pdf/1802.04151.pdf for details.
 
         Args:
           pol : polarization string, Ex. a pseudo-stokes pol 'pI', or a linear pol 'XX'
 
         Returns:
           omega : float, integral of the beam across the sky [steradians]
         """
-        if self.beam_type != 'power':
-            raise ValueError('beam_type must be power')
+        if isinstance(pol, (str, np.str)):
+            pol = uvutils.polstr2num(pol)
+        if pol in [-5, -6, -7, -8]:
+            if self.beam_type != 'power':
```

Comment:
I think users will find it confusing that to get 'xx' or 'yy' you need a power beam but to get a pseudo-stokes beam you need an efield beam. I think we should make it consistent and calculate whatever is needed to the extent possible. I think in general it should work for either type unless the information you need isn't present or has been lost.

Reply:
True. This is no more the case. You can use efield_to_pseudo function to convert from efield to pstokes parameters/power.